akka-http-core/src/main/scala/akka/http/impl/engine/http2/Http2Blueprint.scala

```diff
@@ -179,6 +191,49 @@ private[http] object Http2Blueprint {
   def clientDemux(settings: Http2CommonSettings, masterHttpHeaderParser: HttpHeaderParser): BidiFlow[Http2SubStream, FrameEvent, FrameEvent, Http2SubStream, NotUsed] =
     BidiFlow.fromGraph(new Http2ClientDemux(settings, masterHttpHeaderParser))
 
+  val coalesceWrites: BidiFlow[ByteString, ByteString, ByteString, ByteString, NotUsed] =
```

Comment:
This idea is great, I just want to understand some more about if it's supposed to make a difference for our current throughput benchmarks. In those each "virtual user" emits unary calls sequentially and waits for response for each request before emitting next. We use rather many virtual users (up to 1500). Each virtual user has its own gRPC connection. The request and response messages are small.

Reply:
Right now we send frames one-by-one. A gRPC usually needs at least 3 frames (initial headers, data, trailer headers), so we can reduce the number of system calls by a factor of three if we can emit those in one chunk.

There's a commented out variant that just schedules an AsyncInput which will collect all frames that are the result of the current run (i.e. a response should be sent). That version wouldn't introduce a fixed delay.

The wait time will help if there are multiple responses that should be sent out in about the same time frame but over multiple activations of the GraphInterpreter (responses come in from the user asynchronously from a `mapAsyncUnordered`.

I'll also try the alternative of batching together all pending frames from the demuxer directly without the extra stage.

Reply:
I see, headers, data, trailer explains why we probably saw an improvement also for the "single-threaded unary call clients"

Reply:
> This idea is great, I just want to understand some more about if it's supposed to make a difference for our current throughput benchmarks. In those each "virtual user" emits unary calls sequentially and waits for response for each request before emitting next. We use rather many virtual users (up to 1500). Each virtual user has its own gRPC connection. The request and response messages are small.

I tested again emulating that particular behavior (i.e. concurrent requests == num connections), and indeed in that case, it doesn't seem as if coalescing writes will make a significant difference.

That said, we leave a lot of performance on the table if we don't use a load balancer that bundles up incoming user requests into a smaller number of load-balancer->backend connections.